Flaky `TestConcurrentFetchers_fetchSingle/should_return_an_error_response_if_the_Fetch_request_contains_an_error`

[Logs](https://github.com/grafana/mimir/actions/runs/14212346738/job/39821721618?pr=11084#step:8:93):

```
--- FAIL: TestConcurrentFetchers_fetchSingle (0.08s)
    --- FAIL: TestConcurrentFetchers_fetchSingle/should_return_an_error_response_if_the_Fetch_request_contains_an_error (0.03s)
        logger.go:38: 2025-04-02 05:41:00.823465896 +0000 UTC m=+21.680435290 level info msg starting concurrent fetchers start_offset 0 concurrency 1 bytes_per_fetch_request 2147483647
        logger.go:38: 2025-04-02 05:41:00.823491426 +0000 UTC m=+21.680460841 level info component kafka_client msg producing to a new topic for the first time, fetching metadata to learn its partitions topic test-topic
        logger.go:38: 2025-04-02 05:41:00.823566082 +0000 UTC m=+21.680535467 fetcher 0 level error msg received an error while fetching records; will retry after processing received records (if any) duration 13.515µs start_offset 0 end_offset 214748 asked_records 214748 got_records 0 diff_records 214748 asked_bytes 2147483647 got_bytes 0 diff_bytes 2147483647 first_timestamp  last_timestamp  hwm 0 lso 0 err unknown partition leader
        logger.go:38: 2025-04-02 05:41:00.823494528 +0000 UTC m=+21.680463913 level info component kafka_client msg immediate metadata update triggered why forced load because we are producing to a topic for the first time
        logger.go:38: 2025-04-02 05:41:00.827098806 +0000 UTC m=+21.684068201 level info component kafka_client msg done waiting for metadata for new topic topic test-topic
        logger.go:38: 2025-04-02 05:41:00.827162254 +0000 UTC m=+21.684131649 level info component kafka_client msg initializing producer id
        logger.go:38: 2025-04-02 05:41:00.82741792 +0000 UTC m=+21.684387315 level info component kafka_client msg producer id initialization success id 4457468340882392421 epoch 0
        logger.go:38: 2025-04-02 05:41:00.836472957 +0000 UTC m=+21.693442352 fetcher 0 level error msg received an error while fetching records; will retry after processing received records (if any) duration 606.964µs start_offset 0 end_offset 214748 asked_records 214748 got_records 0 diff_records 214748 asked_bytes 2147483647 got_bytes 0 diff_bytes 2147483647 first_timestamp  last_timestamp  hwm 0 lso 0 err fetch request failed with error: UNKNOWN_SERVER_ERROR: The server experienced an unexpected error when processing the request.
        logger.go:38: 2025-04-02 05:41:00.836568976 +0000 UTC m=+21.693538371 fetcher 0 method concurrentFetcher.fetch.attempt level error msg received an error we're not prepared to handle; this shouldn't happen; please report this as a bug err fetch request failed with error: UNKNOWN_SERVER_ERROR: The server experienced an unexpected error when processing the request.
        fetcher_test.go:1055: 
            	Error Trace:	/__w/mimir/mimir/pkg/storage/ingest/fetcher_test.go:966
            	            				/__w/mimir/mimir/pkg/storage/ingest/fetcher_test.go:1055
            	Error:      	Not equal: 
            	            	expected: 0x1
            	            	actual  : 0x2
            	Test:       	TestConcurrentFetchers_fetchSingle/should_return_an_error_response_if_the_Fetch_request_contains_an_error
        logger.go:38: 2025-04-02 05:41:00.84687486 +0000 UTC m=+21.703844254 level info msg stopped concurrent fetchers last_returned_offset -1
FAIL
FAIL	github.com/grafana/mimir/pkg/storage/ingest	96.989s
```

--- a/pkg/storage/ingest/fetcher.go
+++ b/pkg/storage/ingest/fetcher.go
@@ -297,6 +297,32 @@ func newConcurrentFetchers(
 	fetchBackoffConfig backoff.Config,
 	metrics *readerMetrics,
 ) (*concurrentFetchers, error) {
+	f, err := newConcurrentFetchersWithoutStart(ctx, client, logger, topic, partition, startOffset, maxBufferedBytesLimit, trackCompressedBytes, minBytesWaitTime, offsetReader, startOffsetsReader, fetchBackoffConfig, metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	f.wg.Add(1)
+	go f.start(ctx, startOffset, concurrency)
+
+	return f, nil
+}
+
+func newConcurrentFetchersWithoutStart(
+	ctx context.Context,
+	client *kgo.Client,
+	logger log.Logger,
+	topic string,
+	partition int32,
+	startOffset int64,
+	maxBufferedBytesLimit int32,
+	trackCompressedBytes bool,
+	minBytesWaitTime time.Duration,
+	offsetReader *partitionOffsetClient,
+	startOffsetsReader *genericOffsetReader[int64],
+	fetchBackoffConfig backoff.Config,
+	metrics *readerMetrics,
+) (*concurrentFetchers, error) {
 	if fetchBackoffConfig.MaxBackoff == 0 {
 		// Ensure it's not the zero value, which means we haven't got the backoff config due to a bug.
 		return nil, errors.New("fetchBackoffConfig.MaxBackoff has not been set")
@@ -354,9 +380,6 @@ func newConcurrentFetchers(
 		return nil, fmt.Errorf("failed to find topic ID: %w", err)
 	}
 	f.topicID = topics[topic].ID
-
-	f.wg.Add(1)
-	go f.start(ctx, startOffset, concurrency)
 
 	return f, nil
 }

--- a/pkg/storage/ingest/fetcher.go
+++ b/pkg/storage/ingest/fetcher.go
@@ -53,6 +53,9 @@ type fetcher interface {
 	// record and use the returned context when doing something that is common to all records.
 	PollFetches(context.Context) (kgo.Fetches, context.Context)
 
+	// Start starts the fetcher.
+	Start(ctx context.Context)
+
 	// Stop stops the fetcher.
 	Stop()
 
@@ -265,7 +268,9 @@ type concurrentFetchers struct {
 	orderedFetches chan fetchResult
 
 	lastReturnedOffset int64
-	startOffsets       *genericOffsetReader[int64]
+	startOffset        int64
+	startConcurrency   int
+	startOffsetsReader *genericOffsetReader[int64]
 
 	// fetchBackoffConfig is the config to use for the backoff in case of Fetch errors.
 	// We set it here so that tests can override it run faster.
@@ -289,32 +294,6 @@ func newConcurrentFetchers(
 	partition int32,
 	startOffset int64,
 	concurrency int,
-	maxBufferedBytesLimit int32,
-	trackCompressedBytes bool,
-	minBytesWaitTime time.Duration,
-	offsetReader *partitionOffsetClient,
-	startOffsetsReader *genericOffsetReader[int64],
-	fetchBackoffConfig backoff.Config,
-	metrics *readerMetrics,
-) (*concurrentFetchers, error) {
-	f, err := newConcurrentFetchersWithoutStart(ctx, client, logger, topic, partition, startOffset, maxBufferedBytesLimit, trackCompressedBytes, minBytesWaitTime, offsetReader, startOffsetsReader, fetchBackoffConfig, metrics)
-	if err != nil {
-		return nil, err
-	}
-
-	f.wg.Add(1)
-	go f.start(ctx, startOffset, concurrency)
-
-	return f, nil
-}
-
-func newConcurrentFetchersWithoutStart(
-	ctx context.Context,
-	client *kgo.Client,
-	logger log.Logger,
-	topic string,
-	partition int32,
-	startOffset int64,
 	maxBufferedBytesLimit int32,
 	trackCompressedBytes bool,
 	minBytesWaitTime time.Duration,
@@ -360,7 +339,9 @@ func newConcurrentFetchersWithoutStart(
 		metrics:                 metrics,
 		minBytesWaitTime:        minBytesWaitTime,
 		lastReturnedOffset:      startOffset - 1,
-		startOffsets:            startOffsetsReader,
+		startOffset:             startOffset,
+		startConcurrency:        concurrency,
+		startOffsetsReader:      startOffsetsReader,
 		trackCompressedBytes:    trackCompressedBytes,
 		maxBufferedBytesLimit:   maxBufferedBytesLimit,
 		tracer:                  recordsTracer(),
@@ -396,6 +377,12 @@ func (r *concurrentFetchers) BufferedBytes() int64 {
 
 func (r *concurrentFetchers) EstimatedBytesPerRecord() int64 {
 	return r.estimatedBytesPerRecord.Load()
+}
+
+// Start implements fetcher.
+func (r *concurrentFetchers) Start(ctx context.Context) {
+	r.wg.Add(1)
+	go r.start(ctx, r.startOffset, r.startConcurrency)
 }
 
 // Stop implements fetcher.
@@ -698,7 +685,7 @@ func (r *concurrentFetchers) run(ctx context.Context, wants chan fetchWant, logg
 			// Run a single Fetch request.
 			if res := r.fetchSingle(ctx, w); res.Err != nil {
 				// We got an error. We handle it and then discard this fetch result content.
-				w = handleKafkaFetchErr(res.Err, w, errBackoff, r.startOffsets, r.client, attemptSpan)
+				w = handleKafkaFetchErr(res.Err, w, errBackoff, r.startOffsetsReader, r.client, attemptSpan)
 			} else {
 				// We increase the count of buffered records as soon as we fetch them.
 				r.bufferedFetchedRecords.Add(int64(len(res.Records)))

--- a/pkg/storage/ingest/fetcher_test.go
+++ b/pkg/storage/ingest/fetcher_test.go
@@ -1198,48 +1198,28 @@ func createConcurrentFetchers(ctx context.Context, t *testing.T, client *kgo.Cli
 		return offsetReader.FetchPartitionStartOffset(ctx, partition)
 	}, time.Second, logger)
 
-	var (
-		f   *concurrentFetchers
-		err error
+	f, err := newConcurrentFetchers(
+		ctx,
+		client,
+		logger,
+		topic,
+		partition,
+		startOffset,
+		concurrency,
+		maxInflightBytes,
+		true,        // kfake uses compression and behaves similar to apache kafka
+		time.Second, // same order of magnitude as the real one (defaultMinBytesMaxWaitTime), but faster for tests
+		offsetReader,
+		startOffsetsReader,
+		fastFetchBackoffConfig,
+		&metrics,
 	)
+	require.NoError(t, err)
 
 	if start {
-		f, err = newConcurrentFetchers(
-			ctx,
-			client,
-			logger,
-			topic,
-			partition,
-			startOffset,
-			concurrency,
-			maxInflightBytes,
-			true,        // kfake uses compression and behaves similar to apache kafka
-			time.Second, // same order of magnitude as the real one (defaultMinBytesMaxWaitTime), but faster for tests
-			offsetReader,
-			startOffsetsReader,
-			fastFetchBackoffConfig,
-			&metrics,
-		)
-	} else {
-		f, err = newConcurrentFetchersWithoutStart(
-			ctx,
-			client,
-			logger,
-			topic,
-			partition,
-			startOffset,
-			maxInflightBytes,
-			true,        // kfake uses compression and behaves similar to apache kafka
-			time.Second, // same order of magnitude as the real one (defaultMinBytesMaxWaitTime), but faster for tests
-			offsetReader,
-			startOffsetsReader,
-			fastFetchBackoffConfig,
-			&metrics,
-		)
+		f.Start(ctx)
+		t.Cleanup(f.Stop)
 	}
-
-	require.NoError(t, err)
-	t.Cleanup(f.Stop)
 
 	return f, reg
 }

--- a/pkg/storage/ingest/fetcher_test.go
+++ b/pkg/storage/ingest/fetcher_test.go
@@ -326,7 +326,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// This should not block forever now
 		fetches, fetchCtx := fetchers.PollFetches(ctx)
@@ -350,7 +350,7 @@ func TestConcurrentFetchers(t *testing.T) {
 			produceRecord(ctx, t, client, topicName, partitionID, []byte(fmt.Sprintf("record-%d", i)))
 		}
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		fetches := longPollFetches(fetchers, 5, 2*time.Second)
 		assert.Equal(t, fetches.NumRecords(), 5)
@@ -368,7 +368,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// Produce some records after starting the fetchers
 		for i := 0; i < 3; i++ {
@@ -389,7 +389,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// Produce some records
 		for i := 0; i < 5; i++ {
@@ -440,7 +440,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// Produce some records
 		for i := 0; i < 10; i++ {
@@ -471,7 +471,7 @@ func TestConcurrentFetchers(t *testing.T) {
 				_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 				client := newKafkaProduceClient(t, clusterAddr)
 
-				fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+				fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 				// Produce some records
 				for i := 0; i < 20; i++ {
@@ -507,7 +507,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		lastOffset := produceRecord(ctx, t, client, topicName, partitionID, []byte("last-initial-record"))
 
 		// Start fetchers from the offset after the initial records
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, lastOffset-1, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, lastOffset-1, concurrency, 0, true)
 
 		// Produce some more records
 		for i := 0; i < 3; i++ {
@@ -542,7 +542,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		for round := 0; round < 3; round++ {
 			t.Log("starting round", round)
@@ -587,7 +587,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// Produce enough records to saturate each fetcher.
 		const initiallyProducedRecords = concurrency * 10
@@ -648,7 +648,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		fetchRequestCount := atomic.NewInt64(0)
 		maxRequestedOffset := atomic.NewInt64(-1)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// Produce initial records
 		var producedRecordsBytes [][]byte
@@ -770,7 +770,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
 		client := newKafkaProduceClient(t, clusterAddr)
 
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, 0, true)
 
 		// Produce some records.
 		for i := 0; i < 10; i++ {
@@ -818,7 +818,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		}
 
 		// Create fetchers with tracking of uncompressed bytes
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, maxInflightBytes)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, maxInflightBytes, true)
 
 		// Wait for buffered records to stabilize, we expect that they stabilize because the limit is in effect.
 		waitForStableBufferedRecords(t, fetchers)
@@ -874,7 +874,7 @@ func TestConcurrentFetchers(t *testing.T) {
 		client := newKafkaProduceClient(t, clusterAddr)
 
 		// Create fetchers early to ensure we don't miss any records
-		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, maxInflightBytes)
+		fetchers, _ := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, maxInflightBytes, true)
 
 		// Produce large records
 		largeValue := bytes.Repeat([]byte{'a'}, largeRecordSize)
@@ -943,7 +943,9 @@ func TestConcurrentFetchers_fetchSingle(t *testing.T) {
 			ctx                  = context.Background()
 			cluster, clusterAddr = testkafka.CreateCluster(t, partitionID+1, topic)
 			client               = newKafkaProduceClient(t, clusterAddr)
-			fetchers, reg        = createConcurrentFetchers(ctx, t, client, topic, partitionID, 0, 1, 0)
+
+			// Do not start the fetchers, because we want to be in control of the single Fetch request.
+			fetchers, reg = createConcurrentFetchers(ctx, t, client, topic, partitionID, 0, 1, 0, false)
 		)
 		t.Cleanup(cluster.Close)
 
@@ -1066,7 +1068,9 @@ func TestConcurrentFetchers_parseFetchResponse(t *testing.T) {
 		ctx            = context.Background()
 		_, clusterAddr = testkafka.CreateCluster(t, partitionID+1, topic)
 		client         = newKafkaProduceClient(t, clusterAddr)
-		fetchers, _    = createConcurrentFetchers(ctx, t, client, topic, partitionID, 0, 1, 0)
+
+		// Do not start the fetchers because we want to test a single function in isolation.
+		fetchers, _ = createConcurrentFetchers(ctx, t, client, topic, partitionID, 0, 1, 0, false)
 	)
 
 	t.Run("should return error if the response does not contain any topic", func(t *testing.T) {
@@ -1178,7 +1182,7 @@ func TestFetchResult_Merge(t *testing.T) {
 	})
 }
 
-func createConcurrentFetchers(ctx context.Context, t *testing.T, client *kgo.Client, topic string, partition int32, startOffset int64, concurrency int, maxInflightBytes int32) (*concurrentFetchers, prometheus.Gatherer) {
+func createConcurrentFetchers(ctx context.Context, t *testing.T, client *kgo.Client, topic string, partition int32, startOffset int64, concurrency int, maxInflightBytes int32, start bool) (*concurrentFetchers, prometheus.Gatherer) {
 	logger := testingLogger.WithT(t)
 
 	reg := prometheus.NewPedanticRegistry()
@@ -1194,22 +1198,46 @@ func createConcurrentFetchers(ctx context.Context, t *testing.T, client *kgo.Cli
 		return offsetReader.FetchPartitionStartOffset(ctx, partition)
 	}, time.Second, logger)
 
-	f, err := newConcurrentFetchers(
-		ctx,
-		client,
-		logger,
-		topic,
-		partition,
-		startOffset,
-		concurrency,
-		maxInflightBytes,
-		true,        // kfake uses compression and behaves similar to apache kafka
-		time.Second, // same order of magnitude as the real one (defaultMinBytesMaxWaitTime), but faster for tests
-		offsetReader,
-		startOffsetsReader,
-		fastFetchBackoffConfig,
-		&metrics,
+	var (
+		f   *concurrentFetchers
+		err error
 	)
+
+	if start {
+		f, err = newConcurrentFetchers(
+			ctx,
+			client,
+			logger,
+			topic,
+			partition,
+			startOffset,
+			concurrency,
+			maxInflightBytes,
+			true,        // kfake uses compression and behaves similar to apache kafka
+			time.Second, // same order of magnitude as the real one (defaultMinBytesMaxWaitTime), but faster for tests
+			offsetReader,
+			startOffsetsReader,
+			fastFetchBackoffConfig,
+			&metrics,
+		)
+	} else {
+		f, err = newConcurrentFetchersWithoutStart(
+			ctx,
+			client,
+			logger,
+			topic,
+			partition,
+			startOffset,
+			maxInflightBytes,
+			true,        // kfake uses compression and behaves similar to apache kafka
+			time.Second, // same order of magnitude as the real one (defaultMinBytesMaxWaitTime), but faster for tests
+			offsetReader,
+			startOffsetsReader,
+			fastFetchBackoffConfig,
+			&metrics,
+		)
+	}
+
 	require.NoError(t, err)
 	t.Cleanup(f.Stop)
 

--- a/pkg/storage/ingest/reader.go
+++ b/pkg/storage/ingest/reader.go
@@ -250,7 +250,8 @@ func (r *PartitionReader) start(ctx context.Context) (returnErr error) {
 		if err != nil {
 			return errors.Wrap(err, "creating concurrent fetchers during startup")
 		}
-		r.Start(ctx)
+		f.Start(ctx)
+
 		r.setFetcher(f)
 	} else {
 		// When concurrent fetch is disabled we read records directly from the Kafka client, so we want it

--- a/pkg/storage/ingest/reader.go
+++ b/pkg/storage/ingest/reader.go
@@ -125,7 +125,12 @@ func newPartitionReader(kafkaCfg KafkaConfig, partitionID int32, instanceID stri
 	return r, nil
 }
 
-// Stop implements fetcher
+// Start implements fetcher.
+func (r *PartitionReader) Start(context.Context) {
+	// Given the partition reader has no concurrency it doesn't support starting anything.
+}
+
+// Stop implements fetcher.
 func (r *PartitionReader) Stop() {
 	// Given the partition reader has no concurrency it doesn't support stopping anything.
 }
@@ -245,6 +250,7 @@ func (r *PartitionReader) start(ctx context.Context) (returnErr error) {
 		if err != nil {
 			return errors.Wrap(err, "creating concurrent fetchers during startup")
 		}
+		r.Start(ctx)
 		r.setFetcher(f)
 	} else {
 		// When concurrent fetch is disabled we read records directly from the Kafka client, so we want it

